trin-core/src/utp/stream.rs

```diff
@@ -488,179 +496,249 @@ impl UtpSocket {
         self.send_packet(&mut packet);
     }
 
-    fn handle_packet(&mut self, packet: Packet) {
+    fn handle_packet(&mut self, packet: &Packet, src: Enr) -> anyhow::Result<Option<Packet>> {
         debug!(
             "Handle packet: {:?}. Conn state: {:?}",
             packet.get_type(),
             self.state
         );
 
-        // Update remote window size
-        self.remote_wnd_size = packet.wnd_size();
-        debug!("Remote window size: {}", self.remote_wnd_size);
-
         // Only acknowledge this if this follows the last one, else do it when we advance the send
         // window
         if packet.seq_nr().wrapping_sub(self.ack_nr) == 1 {
             self.ack_nr = packet.seq_nr();
         }
 
+        // Reset connection if connection id doesn't match and this isn't a SYN
+        if packet.get_type() != PacketType::Syn
+            && self.state != SocketState::SynSent
+            && !(packet.connection_id() == self.sender_connection_id
+                || packet.connection_id() == self.receiver_connection_id)
+        {
+            return Ok(Some(self.prepare_reply(packet, PacketType::Reset)));
+        }
+
+        // Update remote window size
+        self.remote_wnd_size = packet.wnd_size();
+        debug!("Remote window size: {}", self.remote_wnd_size);
+
         // Update remote peer's delay between them sending the packet and us receiving it
         let now = now_microseconds();
         self.their_delay = abs_diff(now, packet.timestamp());
         debug!("self.their_delay: {}", self.their_delay);
 
-        match packet.get_type() {
-            PacketType::Data => self.handle_data_packet(packet),
-            PacketType::Fin => self.handle_finalize_packet(),
-            PacketType::State => self.handle_state_packet(packet),
-            PacketType::Reset => unreachable!("Reset should never make it here"),
-            PacketType::Syn => self.handle_syn_packet(packet),
+        match (self.state, packet.get_type()) {
```

Comment:
Not sure how we can improve code readability here because we have a variety of cases. For now, I added some more comments, hope this helps. Imo is not a huge PITA because you get used to it fast  :smile: , maybe we can look more into it in later PR.